Implement kubectl debug profiles: general, baseline, and restricted

sding3 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Implements "general", "baseline", and "restricted" debugging profiles for kubectl debug as specified in KEP-1441.

#### Which issue(s) this PR fixes:

xref https://github.com/kubernetes/kubectl/issues/1108

#### Special notes for your reviewer:

I had picked up where @knight42 had left off in https://github.com/kubernetes/kubernetes/pull/110526

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Added "general", "baseline", and "restricted" debugging profiles for kubectl debug.
```

knight42 committed:
feat(debug): add more profiles

Signed-off-by: Jian Zeng <anonymousknight96@gmail.com>

knight42 committed:
feat(debug): implment serveral debugging profiles

Including `general`, `baseline` and `restricted`.

I plan to add more profiles afterwards, but I'd like to get early
reviews.

Signed-off-by: Jian Zeng <anonymousknight96@gmail.com>

knight42 committed:
test: add some basic tests

Signed-off-by: Jian Zeng <anonymousknight96@gmail.com>

knight42 committed:
chore: add some helper functions

Signed-off-by: Jian Zeng <anonymousknight96@gmail.com>

sding3 committed:
ensure pod copies always get their probes cleared

not wanting probes to be present is something we want
for all the debug profiles; so an easy place to implement
this is at the time of pod copy generation.

sding3 committed:
ensure debug container in pod copy is added before the profile application

The way that the container list modification was defered causes the
debug container to be added after the profile applier runs. We now
make sure to have the container list modification happen before
the profile applier runs.

sding3 committed:
make switch over pod copy, ephemeral, or node more clear

sding3 committed:
use helper functions

added a helper function to modify a container out of a list that
matches the provided container name.

also added a helper function that adds capabilities to container
security.

sding3 committed:
add tests for the debug profiles

sding3 commented:
/cc @verb @knight42

sding3 committed:
document new debugging profiles in command line help text

aimuz commented:
/ok-to-test

sding3 committed:
add file header to profiles_test.go

aimuz commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:
```diff
@@ -654,9 +659,9 @@ func (o *DebugOptions) generatePodCopyWithDebugContainer(pod *corev1.Pod) (*core
 			Name:                     name,
 			TerminationMessagePolicy: corev1.TerminationMessageReadFile,
 		}
-		defer func() {
+		appendDebugContainer = func() {
 			copied.Spec.Containers = append(copied.Spec.Containers, *c)
-		}()
+		}
```
This looks like it changed his behavior

aimuz replied:
Thank you for your response

sding3 replied:
Yes, we need this to ensure the generated debug container is added the the container list prior to the `o.Applier.Apply` occurs on line 692 below or otherwise the generated debug container isn't available to `o.Applier`.

sding3 replied:
Nice idea - done in 861a2a011bbff8c863a252e71c6c4e711d14b871

aimuz reviewed (commented):
Thank you

sding3 commented:
/cc @verb @knight42

aimuz reviewed (commented):
Thanks, looks like no problem

/lgtm

aimuz commented:
/assign @soltysh

Tinche commented:
I was just starting to play around with `kubectl debug` in our newly upgraded clusters and ran into the issue of copied probes making it infeasible. Thank you for working on this ❤️

soltysh commented:
/assign @ardaguclu 
since we've recently spoke about this area it'll be best if you could take it over

ardaguclu reviewed (commented):
Thanks for the PR, profiles looks good to me and left a few comments about it. 

In addition to that maybe we can make two changes to extend profiling capability even further. I have already PR for that https://github.com/kubernetes/kubernetes/pull/111457 but it would be much better if we can handle it in this PR as well as new profiles.
* Currently, we don't support external developer's custom profiles. We can also support this by exporting `applier` field https://github.com/kubernetes/kubernetes/blob/a9593d634c6a053848413e600dadbf974627515f/staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go#L229 and setting it only when it is nil. What do you think?
*  Second one is, there is no way of passing variables to profiles. What if user want keeping labels via `--keep-readiness` flag. We might pass an external argument. Currently, it will not be used, but it will provide a way of extensibility.

What do you think about these changes?

sding3 committed:
remove URL to KEP from help text

sding3 committed:
move probe removal to the profiles

sding3 committed:
remove mustNewProfileApplier in tests

sding3 committed:
remove extra whiteline from import block

sding3 commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -49,11 +62,195 @@ func profileLegacy(pod *corev1.Pod, containerName string, target runtime.Object)
-		pod.Spec.HostIPC = true
-		pod.Spec.HostNetwork = true
-		pod.Spec.HostPID = true
+		setHostNamespace(pod, true)
```
well - it does get called in quite a number of places so I prefer to keep it. But maybe we should give this a better name?

ardaguclu replied:
This function includes 3 lines. Since in Go readability has top priority, it would be better just explicitly set in the place instead calling this function.

sding3 committed:
remove isPodCopy helper func

sding3 commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -49,11 +62,195 @@ func profileLegacy(pod *corev1.Pod, containerName string, target runtime.Object)
 	}
 }
+
+func isPodCopy(pod *corev1.Pod, target runtime.Object) bool {
```
Sounds good - done in 72aacf6e0afe90e1bd9dbc141cb98175177e1316.

sding3 commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -49,11 +62,195 @@ func profileLegacy(pod *corev1.Pod, containerName string, target runtime.Object)
+type debugStyle int
+
+const (
+	styleEphemeral debugStyle = iota
```
I think the consistent prefix makes it clearer at call sites. As another example, all the profile name strings are prefixed with the sub-string "profile".

Because all the profiles have to do this exact branching logic based on the debug technique, I think it's good to have this to slightly reduce boilerplate code, improve readability, and improve consistency in profile's implementations.

ardaguclu replied:
I'd suggest;

```go
Ephemeral debugType = iot
```

sding3 replied:
Dropped in 98878e9a835d52da57cc439907a25b560117446b

sding3 commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -49,11 +62,195 @@ func profileLegacy(pod *corev1.Pod, containerName string, target runtime.Object)
+	case *corev1.Node:
+		return styleNode
+	}
+	return debugStyle(-1) // unknown
```
Given that the result is consumed in a `switch` statement, returning multiple values including an `error` values makes the downstream consuming semantics a little less ergonomic. It would also lead to a `default` case that would not be reachable or a trailing return statement after the switch which is not reachable.

ardaguclu replied:
This style looks not Go style. It should return error, if case is not found.

sding3 committed:
switch baselineProfile to using the modifyEphemeralContainer helper

sding3 commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -49,11 +62,195 @@ func profileLegacy(pod *corev1.Pod, containerName string, target runtime.Object)
+		})
+
+	case styleEphemeral:
+		for i := range pod.Spec.EphemeralContainers {
```
Oops - missed it there; switched to using `modifyEphemeralContainer` via 5800849a8ac76e9604e9d13c31416d0d236baa9b

sding3 committed:
rename addCap to addCapability, and don't do deep copy

sding3 commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -49,11 +62,195 @@ func profileLegacy(pod *corev1.Pod, containerName string, target runtime.Object)
+		}
+	}
+
+	ss := s.DeepCopy()
```
Because most profiles will need to do this, it helps with readability and consistency to provide a helper function. It's very easy to mess this up and induce a panic if we expect each profile to implement this due to `nil` pointers.

We don't technically need to do the deep copy, I was only doing to have it so that the input argument `s` is not mutated. I've removed the deep copy.

To further improve the readability at call sites, I've renamed this from `addCap` to `addCapability`.

This is all done in 1b9e3ad05be421f0d3e0949e03f2d198a3f128ac

sding3 commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -49,11 +62,195 @@ func profileLegacy(pod *corev1.Pod, containerName string, target runtime.Object)
+}
+
+// modifyContainer performs m against a container from cs which has the name of containerName.
+func modifyContainer(cs []corev1.Container, containerName string, m func(*corev1.Container)) {
```
I think it would be better to keep this given that most, if not all, profiles currently make use of this boilerplate code.

The name `modifyContainer` makes it clear at call sites and this does.

sding3 committed:
fix godoc on modifyEphemeralContainer

sding3 commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -49,11 +62,195 @@ func profileLegacy(pod *corev1.Pod, containerName string, target runtime.Object)
+}
+
+// modifyContainer performs m against a container from cs which has the name of containerName.
+func modifyEphemeralContainer(cs []corev1.EphemeralContainer, containerName string, m func(*corev1.EphemeralContainer)) {
```
I think it would be better to keep this given that most, if not all, profiles currently make use of this boilerplate code.

I had a typo on the godoc here, I've fixed in f73b4800261cdb70fa02e6d883a5eb3edf456c76

sding3 commented:
> Thanks for the PR, profiles looks good to me and left a few comments about it.
> 
> In addition to that maybe we can make two changes to extend profiling capability even further. I have already PR for that #111457 but it would be much better if we can handle it in this PR as well as new profiles.
> 
>     * Currently, we don't support external developer's custom profiles. We can also support this by exporting `applier` field https://github.com/kubernetes/kubernetes/blob/a9593d634c6a053848413e600dadbf974627515f/staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go#L229
>        and setting it only when it is nil. What do you think?
> 
>     * Second one is, there is no way of passing variables to profiles. What if user want keeping labels via `--keep-readiness` flag. We might pass an external argument. Currently, it will not be used, but it will provide a way of extensibility.
> 
> 
> What do you think about these changes?

@ardaguclu - thank you for review and sharing your thoughts! I've responded to your comments. I am also amenable to implementing further enhancement to the extensibility of debug files as you've suggested. However, I'd like to do that in a separate/subsequent PR if that's alright with you.

sding3 committed:
export DebugOptions.Applier for extensibility

sding3 commented:
@ardaguclu - on second thought, I do think that it makes sense to go ahead and export `Applier`, I have done that via 480c094675e6327b80422489e3edeeac6f5a5314

sding3 committed:
fix unit test

sding3 committed:
fix spelling on overriden

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:
```diff
@@ -543,6 +543,8 @@ func (o *DebugOptions) generateDebugContainer(pod *corev1.Pod) (*corev1.Pod, *co
 		return nil, nil, err
 	}
 
+	ec = &copied.Spec.EphemeralContainers[len(copied.Spec.EphemeralContainers)-1]
```
Is this change related to new debug profiles or is it fixing something?. If it fixes a bug, it is better to fix that independent from this PR.

sding3 replied:
It is related to new debug profiles, because the only pre-existing `legacy` profile doesn't update `EphemeralContainers`.

sding3 replied:
Sounds good, I've made https://github.com/kubernetes/kubernetes/issues/115285 to track that. I'd be happy to circle back around on this in a future MR.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -49,11 +61,195 @@ func profileLegacy(pod *corev1.Pod, containerName string, target runtime.Object)
+		// For copy of pod: empty securityContext; sets shareProcessNamespace
+		pod.Spec.SecurityContext = nil
+		pod.Spec.ShareProcessNamespace = pointer.BoolPtr(true)
+		setHostNamespace(pod, false)
```
I agree that all these functions `setHostNamespace`, `modifyContainer`, `removeProbes` preventing a lot of duplications and it causes less code and appreciate the effort. However, I think duplication is not a bad thing in this case. Because when someone in the future read the code, it immediately should understand what all these profiles are doing.

For example, what we are losing by this;

```go
// For copy of pod: empty securityContext; sets shareProcessNamespace
pod.Spec.SecurityContext = nil
pod.Spec.ShareProcessNamespace = pointer.BoolPtr(true)
pod.Spec.HostNetwork = false
pod.Spec.HostPID = false
pod.Spec.HostIPC = false
for i, c := range pod.Spec.Containers {
   pod.Spec.Containers[i].SecurityContext =  &corev1.SecurityContext{
			Capabilities: &corev1.Capabilities{
				Add: []corev1.Capability{c},
			},
		}
  pod.Spec.Containers[i].LivenessProbe = nil
  pod.Spec.Containers[i].ReadinessProbe = nil
}

```

verb replied:
imo, for profiles we want to prioritize composability. We have a lot of profiles that differ in only minor ways. From a readability perspective, I think what's interesting is not what each profile does (which is simple) but how it differs from the other profiles (which is complicated). Helpers make it easy to see how profiles differ.

I generally agree that duplication is not necessarily a bad thing, and it's not duplication that I'm trying to avoid. It was my hope that careful helper naming would improve readability, like:

```
general:
  RemoveLabelsAndProbes()
  case styleNode:
    ClearSecurityContext()
    UseHostNamespaces()
    MountRootPartition()
  case stylePodCopy:
    AllowProcessTracing()
    ShareProcessNamespace()
  case styleEphemeral:
    AllowProcessTracing()

baseline:
  RemoveLabelsAndProbes()
  case styleNode:
    ClearSecurityContext()
  case stylePodCopy:
    ClearSecurityContext()
    ShareProcessNamespace()
  case styleEphemeral:
    ClearSecurityContext()

restricted:
  DisallowRoot()
  DropCapabilities()
  case stylePodCopy:
    ShareProcessNamespace()
  case styleEphemeral:
    ClearSecurityContext()
```

What do you think?

ardaguclu replied:
I think you are right, in order to increase readability further, with good naming we can apply your suggestion

sding3 replied:
Good idea, implemented in 4472c8f6fbdb5d3357fdbaedfb6fcd235a701732

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles_test.go:
```diff
@@ -0,0 +1,429 @@
+	},
+}
+
+func TestGeneralProfile(t *testing.T) {
```
Thanks for all these unit tests.

sding3 committed:
remove debugStyle facilities

sding3 committed:
inline setHostNamespace helper func

sding3 committed:
remove modifyContainer, modifyEphemeralContainer, and remove probes

their logic have been in-lined at call sites

sding3 committed:
remove DebugApplierFunc convenience facility

sding3 committed:
fix baseline profile implementation

it shouldn't have SYS_PTRACE base on
https://github.com/kubernetes/enhancements/tree/master/keps/sig-cli/1441-kubectl-debug#profile-baseline

sding3 committed:
remove addCapability helper, in-lining at call sites

ardaguclu commented:
/test pull-kubernetes-unit

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -49,11 +61,167 @@ func profileLegacy(pod *corev1.Pod, containerName string, target runtime.Object)
+	case *corev1.Pod:
+		if t == pod {
+			// For ephemeral container: sets SYS_PTRACE in ephemeral container
+			for i, c := range pod.Spec.EphemeralContainers {
```
That would be more simplified. Don't we manage debug container whose name is `containerName` which we are looking for and we know that this debug container does not contain any SecurityContext?

```go
			if c.Name != containerName {
				continue
			}
			// For ephemeral container: sets SYS_PTRACE in ephemeral container
			pod.Spec.EphemeralContainers[i].SecurityContext = &corev1.SecurityContext{
				Capabilities: &corev1.Capabilities{
					Add: []corev1.Capability{"SYS_PTRACE"},
				},
			}
```

ardaguclu replied:
and we can do the same all the other profiles.

sding3 replied:
Yeah, you're right. I have pushed the code update in ad29d22cb926929d8bb51f3b12cabdd5694711ea.

However, the reason why I did this as well as explicitly set pod's HostNework, HostPID and HostIPC fields to false in the case of node debugging is that I wanted to avoid a coupling between:

1. the way that the debug container or debug pod (in the case of node debugging) is constructed by the ProfilesApplier's callers, and
2. the ProfileApplier implementations themselves

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -49,11 +61,167 @@ func profileLegacy(pod *corev1.Pod, containerName string, target runtime.Object)
+			}
+		} else {
+			// For copy of pod: sets SYS_PTRACE in debugging container, sets shareProcessNamespace
+			pod.Spec.ShareProcessNamespace = pointer.BoolPtr(true)
```
`BoolPtr` is deprecated, it is better to use `Bool`

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -49,11 +61,167 @@ func profileLegacy(pod *corev1.Pod, containerName string, target runtime.Object)
+		} else {
+			// For copy of pod: sets SYS_PTRACE in debugging container, sets shareProcessNamespace
+			pod.Spec.ShareProcessNamespace = pointer.BoolPtr(true)
+			for i, c := range pod.Spec.Containers {
```
Maybe to get rid of `else if` statements, we can change this to something like tha?;

```go
				if c.Name != containerName {
					continue
				}
				
				pod.Spec.Containers[i].LivenessProbe = nil
				pod.Spec.Containers[i].ReadinessProbe = nil
				
				if c.SecurityContext == nil {
					pod.Spec.Containers[i].SecurityContext = &corev1.SecurityContext{
						Capabilities: &corev1.Capabilities{
							Add: []corev1.Capability{"SYS_PTRACE"},
						},
					}
					continue
				}
				
				if c.SecurityContext.Capabilities == nil {
					pod.Spec.Containers[i].SecurityContext.Capabilities = &corev1.Capabilities{
						Add: []corev1.Capability{"SYS_PTRACE"},
					}
					continue
				}
				
				pod.Spec.Containers[i].SecurityContext.Capabilities.Add = append(c.SecurityContext.Capabilities.Add, "SYS_PTRACE")
```

ardaguclu replied:
But I'd like to note that, code snippet above clears probes only for debug container and not others. If we need to clear probes for all containers. We need to change that.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -49,11 +61,167 @@ func profileLegacy(pod *corev1.Pod, containerName string, target runtime.Object)
+		if t == pod {
+			// For ephemeral container: empty securityContext
+			for i, c := range pod.Spec.EphemeralContainers {
+				if c.Name == containerName {
```
Returning early is recommended way in Go;

```go
if c.Name != containerName{
continue
}
pod.Spec.EphemeralContainers[i].SecurityContext = nil
```

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -49,11 +61,167 @@ func profileLegacy(pod *corev1.Pod, containerName string, target runtime.Object)
+			// For copy of pod: empty securityContext; sets shareProcessNamespace
+			pod.Spec.SecurityContext = nil
+			pod.Spec.ShareProcessNamespace = pointer.BoolPtr(true)
+			pod.Spec.HostNetwork = false
```
Is this necessary?. If we don't set anything, isn't this accepted as `false`?

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -49,11 +61,167 @@ func profileLegacy(pod *corev1.Pod, containerName string, target runtime.Object)
+				if c.Name == containerName {
+					pod.Spec.Containers[i].SecurityContext = nil
+				}
+				pod.Spec.Containers[i].LivenessProbe = nil
```
Is it correct to clear probes for all containers in the pod. Shouldn't we do it only for debug container?

ardaguclu replied:
As far as I can see in here https://github.com/kubernetes/kubernetes/pull/110526/files#diff-789890b76929281f2d5ca2c895ae5294bb6ab24fda7024b31c1db46f452c7c9dR185, probes are cleared for only debug container

sding3 replied:
I think striping every container's probes in the pod copy work better, because in a debug session where you might do something to the network namespace of the pod that causes every container in the pod no longer able to receive or respond to liveness probes, we wouldn't want all the containers to get inadvertently bounced in such a debugging scenario and cause inconvenience. But I'll defer to @verb

verb replied:
The original motivation here was that the probes were causing the pod to be killed in the middle of a debugging session.

Since we're stripping the labels so that the pod won't receive traffic it makes sense to strip the probes as well.

At least that's what I remember. :grin:

sding3 committed:
address Arda's code review comments

1 use Bool instead of BoolPtr (now deprecated)
2 tweak for loop to continue when container name is not what we expect
3 use our knowledge on how the debug container is generated to simplify
  our modification to the security context
4 use our knowledge on how the pod for node debugging is generated to no
  longer explicit set pod's HostNework, HostPID and HostIPC fields to
  false

sding3 commented:
/retest-required

ardaguclu reviewed (commented):
Aside from the unclear part of `liveleness` and `readiness` probes should be cleared only for the debug container or for all of them, this PR looks good to me.

As the KEP owner, I'll defer to @verb for another check.

Thanks for your effort @sding3

ardaguclu commented:
@sding3 in a followup PR or in this PR(whichever you prefer), it would be good to add integration tests for all these profiles in here https://github.com/kubernetes/kubernetes/blob/30262e9b14e2369528fd27241adb16fae02f684d/test/cmd/debug.sh#L21

verb commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -49,11 +61,154 @@ func profileLegacy(pod *corev1.Pod, containerName string, target runtime.Object)
+
+func (p *generalProfile) Apply(pod *corev1.Pod, containerName string, target runtime.Object) error {
+	switch t := target.(type) {
+	case *corev1.Pod:
```
There's going to be a lot of duplication in the profiles, so I recommend creating some helper functions to make this easier to read.

I had imagined something like this:

```
switch t:= target.(type) {
case *corev1.Pod:
  setPtrace(pod, containerName)
  stripAllProbes(pod)
case *corev1.Node:
  SetHostNamespaces(pod)
}
```

What do you think?

sding3 replied:
I think that makes sense - what do you think of c10aec8b092ad1428deb05f76697466348d27e20

verb commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:
```diff
@@ -123,6 +123,7 @@ type DebugOptions struct {
 	TargetContainer string
 	TTY             bool
 	Profile         string
+	Applier         ProfileApplier
```
could you help me catch up on why this is now exported?

sding3 replied:
Arda suggested this during review of this MR:

![image](https://user-images.githubusercontent.com/18542826/214217071-ee3f79df-966a-4bf0-ac9a-04e0cc9c0ecb.png)

ardaguclu replied:
This is basically for the other CLI's like `oc` to use `kubectl debug`. `oc` has it's own custom profiles and it will set `Applier` with custom profiles.

verb reviewed (commented):
Thanks for picking up this effort! :heart: 

There's a bit too much history in this PR for me to read entirely, so I apologize if I'm repeating or contradicting past comments: I'm just trying to catch up.

sding3 committed:
remove tricky defer in generatePodCopyWithDebugContainer

sding3 committed:
provide helper functions to make debug profiles more readable

sding3 committed:
add note to remind people about updating --profile's help text when adding new profiles

sding3 commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:
```diff
@@ -182,7 +181,7 @@ func addDebugFlags(cmd *cobra.Command, opt *DebugOptions) {
 	cmd.Flags().StringVar(&opt.TargetContainer, "target", "", i18n.T("When using an ephemeral container, target processes in this container name."))
 	cmd.Flags().BoolVarP(&opt.TTY, "tty", "t", opt.TTY, i18n.T("Allocate a TTY for the debugging container."))
-	cmd.Flags().StringVar(&opt.Profile, "profile", ProfileLegacy, i18n.T("Debugging profile."))
+	cmd.Flags().StringVar(&opt.Profile, "profile", ProfileLegacy, i18n.T(`Debugging profile. Options are "legacy", "general", "baseline", or "restricted".`))
```
I think that would be useful, but I think given that the KEP only specified a only handful of profiles - I think we should be okay? In any case, I think we should defer adding a new flag to that prints out the list of profiles and exits to a future MR if possible.

For now, I've added a code comment to remind folks in 77f8148289680fd3578945c8b14c6594d0853f2f

sding3 reviewed (commented):
@verb - thank you for reviewing. I've responded to your comments. Let me know what you think.

sding3 commented:
/retest-required

sding3 commented:
> @sding3 in a followup PR or in this PR(whichever you prefer), it would be good to add integration tests for all these profiles in here
> 
> https://github.com/kubernetes/kubernetes/blob/30262e9b14e2369528fd27241adb16fae02f684d/test/cmd/debug.sh#L21

@ardaguclu - sounds good. I created https://github.com/kubernetes/kubernetes/issues/115287 to track the future MR

sding3 committed:
Implement helper functions with names that improve readability

ardaguclu commented:
/triage accepted
/priority backlog

I think, we need to squash all these commits either by @sding3 or tide/merge-method-squash

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -21,39 +21,254 @@ import (
+	case *corev1.Node:
+		return styleNode, nil
+	}
+	return debugStyle(-1), fmt.Errorf("objects of type %T are not supported", target)
```
I know I said this multiple times, but this `debugStyle(-1)` is not gopher way :)

sding3 replied:
Oops - replaced it with `styleUnsupported` in [2111c5c](https://github.com/kubernetes/kubernetes/pull/114280/commits/2111c5cc670aff0d4a0a19285833bfb603f48188)

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -21,39 +21,254 @@ import (
+
+// modifyContainer performs the modifier function m against a container from cs
+// which has the name of containerName and the type satisfying the mask.
+func modifyContainer(pod *corev1.Pod, containerName string, mask ContainerType, m func(*corev1.Container)) {
```
I see that this function does what is intended, but it looks to me too clever than simpler.

sding3 replied:
Yeah, the bit mask is tricky, but I took the idea from [here](https://github.com/kubernetes/kubernetes/blob/6ce055d62d8608842db18901c948c6982c74a164/pkg/api/pod/util.go#L61) 😀

I think most programmers should be familiar with bit masks because the technique is prevalent in the construction of linux syscall arguments as well as certain places in the go std lib, [io/fs.FileMode](https://pkg.go.dev/io/fs#FileMode) for example.

sding3 replied:
I don't think we can import that dependency based on kubectl's [Contribution Requirement](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/kubectl#contribution-requirements).

The `VisitContainers` utility function works with init containers, which we don't currently need. Since what we have satisfies our needs already and what we have is currently unexported, I think we should keep things as they are.

verb replied:
Asking about this on slack, @ardaguclu pointed out that there's already a [podutils](https://github.com/kubernetes/kubernetes/blob/43e8ebbbcd3f57d18d8151efb6242f88a763b06d/staging/src/k8s.io/kubectl/pkg/util/podutils/podutils.go#L60) and suggested copying the `VisitContainers` implementation there.

Do you mind doing the copy and updating profiles to use `VisitContainers`?

Ref: https://kubernetes.slack.com/archives/C2GL57FJ4/p1675803799358839

sding3 replied:
Updated in 5db0224fac026f01fcd5a8ffe166da04b6b5262e

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -21,39 +21,254 @@ import (
+	return nil
+}
+
+func (p *restrictedProfile) Apply(pod *corev1.Pod, containerName string, target runtime.Object) error {
```
These `Apply` functions looks really nice

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -21,39 +21,254 @@ import (
+}
+
+// ContainerType signifies container type
+type ContainerType int
```
We have already `debugStyle`, I think there is no need additional type field?

sding3 replied:
It would be nice to use one concept with both, but I'm afraid that `ContainerType` and `debugStyle` might grow in ways that we may not want to mix the two concepts.

For example, a future debug profile might want to make modifications to init containers when debugging by pod copy. So it needs to add a new `ContainerType` of `InitContainer`. Conversely, a future debugging methodology might target a new runtime object type that is not a container type, which would require adding a new value of `debugStyle`.

sding3 committed:
add styleUnsupported to replace debugStyle(-1)

sding3 committed:
fix godoc on modifyContainer

ardaguclu commented:
Latest version of this PR looks good to me.

For the final approval and/or review, I'm deferring to @verb

verb reviewed (commented):
A couple of questions and minor comments, but I think this looks really good.

I'll circle back and check on the contents of the profiles if I get time, but I'm really happy about how the Applier turned out. Thanks for your hard work here!

verb commented:
> I think, we need to squash all these commits either by @sding3 or tide/merge-method-squash

+1, I think it's time to squash

sding3 commented:
/label tide/merge-method-squash

sding3 committed:
drop style prefix from debugStyle values

sding3 commented:
> A couple of questions and minor comments, but I think this looks really good.
> 
> I'll circle back and check on the contents of the profiles if I get time, but I'm really happy about how the Applier turned out. Thanks for your hard work here!

@verb - I had just responded to your comments, could you PTAL again?

sding3 committed:
put VisitContainers in podutils & use that from debug

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -257,21 +248,12 @@ func AllowProcessTracing(p *corev1.Pod, containerName string, mask ContainerType
-				continue
-			}
-			m(&pod.Spec.Containers[i])
+func modifyContainer(pod *corev1.Pod, containerName string, mask podutils.ContainerType, m func(*corev1.Container)) {
```
In my opinion, there is no need this function. Every caller function can directly pass their Visitor functions(whose interface is `func(c *corev1.Container, _ podutils.ContainerType)`).

Only benefit of this function looks like filtering by `containerName`. I think, we can also do this simply in caller functions.

What do you think @sding3  @verb ?

sding3 replied:
I think it would complicates the call sites, as the all call sites would now have to:
1. perform name filtering,
2. take in a `podutils.ContainerType` in the `VisitFunc` which will be unused since `podutils.VisitContainer` already does type filtering, and 
3. be very careful about whether their `VisitFunc` returns `true` or `false`

ardaguclu replied:
> I think it would complicate the call sites, as all call sites would now have to:
> 
> 1. perform name filtering,

Name filtering is one line of check
> 2. take in a `podutils.ContainerType` in the `VisitFunc` which will be unused since `podutils.VisitContainer` already does type filtering, and

Isn't the benefit of using `VisitContainer` just does not have to maintain `containerType` in here?
> 3. be very careful about whether their `VisitFunc` returns `true` or `false`

Does `modifyContainer` function handle that case already?

My main concern is we are passing referenced type objects into visitor but these referenced type objects are also being passed to outer `m` function. That makes difficult to follow the execution path if there is any bug occurred in the future.

sding3 replied:
> Name filtering is one line of check

It would actually be 3 lines and the additional mental acrobatics to decide whether to return `true` or `false`:

```go
         if c.Name != containerName {
             // return true or false here?
         }
         // do modification
         // return true or false here?
```

By allowing call sites to decide how they want to filter the container name, we now have also introduce the possibility for another variant where the call site uses:

```go
         if c.Name == containerName {
             // do modification
             // return true or false here?
         }
         // return true or false here?
```

> Does modifyContainer function handle that case already?

Yes, modifyContainer currently handles the returning of the correct boolean value.

With all that said, I'm fine with in-lining all the calls if @verb also request so.

verb replied:
I agree with @ardaguclu that this would be better inlined. It's true that it complicates the call site a bit (with return true vs false), but it's a common pattern and also simplifies by shortening the call stack.

Modifying a single container by name would look something like:
```
if c.Name != containerName {
  return true
}

// do modification
return false
```

modifying all containers would be:
```
// do modification
return true
```

sding3 replied:
Sounds good - updated in 1624f5f96dc88e85d5f70f8615685e199d3de93a

sding3 committed:
cite source for ContainerType and VisitContainers

sding3 committed:
pull in AllContainers ContainerType value

sding3 committed:
have VisitContainer take pod spec rather than pod

sding3 committed:
in-line modifyContainer

sding3 commented on staging/src/k8s.io/kubectl/pkg/util/podutils/podutils.go:
```diff
@@ -186,3 +186,48 @@ func maxContainerRestarts(pod *corev1.Pod) int {
 	}
 	return maxRestarts
 }
+
```
Sounds good - added in 88d19f4b5aa505694443c724c6c0aff492af8923

I also noticed that my `VisitContainers` was taking a `*pod` instead of a `*podspec` and I've updagted that in 9980209146c0c5059b26f6fcccb765d7d263a8b9 so that it's now consistent with what's in k/k

verb replied:
Thanks for noticing the inconsistency!

sding3 commented on staging/src/k8s.io/kubectl/pkg/util/podutils/podutils.go:
```diff
@@ -186,3 +186,48 @@ func maxContainerRestarts(pod *corev1.Pod) int {
+// ContainerType signifies container type
+type ContainerType int
+
+const (
```
Sounds good - done in d917ea711637a8519ee2ac780c0f9256f682969e

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -21,39 +21,246 @@ import (
+	}
+}
+
+type debugStyle int
```
nit: It would be better to move constants at the beginning of this file.

sding3 replied:
good idea - done in 7f7ef70651b9d81cf479647805019e0aec3e484e

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -21,39 +21,246 @@ import (
+
+// RemoveLabelsAndProbes removes labels from the pod and remove probes
+// from all containers of the pod.
+func RemoveLabelsAndProbes(p *corev1.Pod) {
```
nit: maybe we can unexport all these helper functions, if there isn't any explicit reason

verb replied:
+1

sding3 replied:
sounds good - done in 98f8ff5a3ef6516128987798111313f32ffc7b41

ardaguclu reviewed (commented):
I quickly tested these profiles on Kind cluster and did not encounter any problem. I left a few unblocker comments can be easily done in followup PR. It would be great if we can also add integration tests in a followup PR too.

Thanks for all your effort @sding3. 

/approve

ardaguclu commented:
@verb could you PTAL for the last tag?

verb commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profile_applier.go:
```diff
@@ -35,15 +45,14 @@ type ProfileApplier interface {
-// applierFunc is a function that applies a profile to a container in the pod.
-type applierFunc func(pod *corev1.Pod, containerName string, target runtime.Object) error
-
-func (f applierFunc) Apply(pod *corev1.Pod, containerName string, target runtime.Object) error {
```
nit: Should this whole file just move into profiles.go at this point?

verb commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -21,39 +21,246 @@ import (
+}
+
+// ClearSecurityContext clears the security context for the container.
+func ClearSecurityContext(p *corev1.Pod, containerName string, mask podutils.ContainerType) {
```
nit: mask is unnecessary. Since we're searching by unique name it can always be `Containers|EphemeralContainers` (or even `AllContainers`: I'd prioritize readability over optimization for this infrequently used loop)

Same for all of the other helpers that take a mask.

sding3 replied:
Ah, I didn't know that we could unambiguously identify by the name alone, simplified in d3583dcd60bc6d5ec2968c2e9852334672ba5852

verb commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -21,39 +21,246 @@ import (
+		if c.Name != containerName {
+			return true
+		}
+		if c.SecurityContext == nil {
```
It's a little difficult to see that these all do the same thing, and I think we probably want to overwrite everything anyway if the goal is to drop them all. What about this?

```
if c.SecurityContext == nil {
  c.SecurityContext = &corev1.SecurityContext{}
}
c.SecurityContext.Capabilities = &corev1.Capabilities{
  Drop: []corev1.Capability{"ALL"},
}
return false
```

sding3 replied:
good idea - done in f337511f064d7e28fa0bf7ab4b27a3af17d178d9

verb commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:
```diff
@@ -21,39 +21,246 @@ import (
+		}
+		if c.SecurityContext.Capabilities == nil {
+			c.SecurityContext.Capabilities = &corev1.Capabilities{
+				Drop: []corev1.Capability{"SYS_PTRACE"},
```
Should this be `Capabilities.Add` instead of `Capabilities.Drop`?  Recommend the same as above to avoid this sort of bug:

```
if c.SecurityContext == nil {
  c.SecurityContext = &corev1.SecurityContext{}
}
if c.SecurityContext.Capabilities == nil {
  c.SecurityContext.Capabilities = &corev1.Capabilities{}
}
c.SecurityContext.Capabilities.Add = append(c.SecurityContext.Capabilities.Add, "SYS_PTRACE")
return false
```

Could you add a test case for this?

sding3 replied:
nice catch - fixed and added a test in f604758c84dbf2e9c068455188564a8aaebc7d69

verb reviewed (commented):
+1, thanks for your dedication here! I know lots of people are eagerly awaiting this feature.

I added a few nits and there's an issue with the capabilities helpers that's worth addressing now. Feel free to take any of the nits as follow-up, I'm happy to lgtm this once the capability issue is addressed.

sding3 committed:
unexport helper funcs

sding3 committed:
put debugStyle at top of file

sding3 committed:
merge profile_applier.go into profile.go

sding3 committed:
tweak dropCapabilities

sding3 committed:
fix allowProcessTracing & add a test for it

sding3 committed:
drop mask param from help funcs, since we can already unambiguous identify the container by name

sding3 commented:
> I quickly tested these profiles on Kind cluster and did not encounter any problem. I left a few unblocker comments can be easily done in followup PR. It would be great if we can also add integration tests in a followup PR too.

Yep, that will be my top priority after this PR.

verb reviewed (approved):
/lgtm

sding3 committed:
fix grammar in code comment

sding3 commented:
@verb - sorry I pushed a grammar fix in d760b0c6a7111046caff8851d1bd8e792e338bfe after you already tagged - could you retag? Thank you.

verb reviewed (approved):
/lgtm

hariskmohd commented:
Is this feature available now or when can we expect this to come?
